include/RAJA/policy/cuda/reduce.hpp

```diff
@@ -94,12 +167,220 @@ __device__ __forceinline__ T shfl_xor(T var, int laneMask)
   Tunion.var = var;
 
   for (int i = 0; i < int_sizeof_T; ++i) {
-    Tunion.arr[i] = __shfl_xor(Tunion.arr[i], laneMask);
+#if (__CUDACC_VER_MAJOR__ >= 9)
+    Tunion.arr[i] = ::__shfl_sync(Tunion.arr[i], srcLane);
+#else
+    Tunion.arr[i] = ::__shfl(Tunion.arr[i], srcLane);
+#endif
   }
   return Tunion.var;
 }
 
-}  // end HIDDEN namespace for helper functions
+
+}  // namespace internal
+
+//! Object that manages pinned memory buffers for reduction results
+//  use one per reducer object
+template <typename T>
+class PinnedTally
+{
+public:
+  //! Object put in Pinned memory with value and pointer to next Node
+  struct Node {
+    Node* next;
+    T value;
+  };
+  //! Object per stream to keep track of pinned memory nodes
+  struct StreamNode {
+    StreamNode* next;
+    cudaStream_t stream;
+    Node* node_list;
+  };
+  
+  //! Iterator over streams used by reducer
+  class StreamIterator {
+  public:
+    StreamIterator() = delete;
+    
+    StreamIterator(StreamNode* sn)
+      : m_sn(sn)
+    {
+    }
+    
+    const StreamIterator& operator++()
+    {
+      m_sn = m_sn->next;
+      return *this;
+    }
+    
+    StreamIterator operator++(int)
+    {
+      StreamIterator ret = *this;
+      this->operator++();
+      return ret;
+    }
+    
+    cudaStream_t& operator*()
+    {
+      return m_sn->stream;
+    }
+    
+    bool operator==(const StreamIterator& rhs) const
+    {
+      return m_sn == rhs.m_sn;
+    }
+    
+    bool operator!=(const StreamIterator& rhs) const
+    {
+      return !this->operator==(rhs);
+    }
+    
+  private:
+    StreamNode* m_sn;
+  };
+  
+  //! Iterator over all values generated by reducer
+  class StreamNodeIterator {
+  public:
+    StreamNodeIterator() = delete;
+    
+    StreamNodeIterator(StreamNode* sn, Node* n)
+      : m_sn(sn), m_n(n)
+    {
+    }
+    
+    const StreamNodeIterator& operator++()
+    {
+      if (m_n->next) {
+        m_n = m_n->next;
+      } else if (m_sn->next) {
+        m_sn = m_sn->next;
+        m_n = m_sn->node_list;
+      } else {
+        m_sn = nullptr;
+        m_n = nullptr;
+      }
+      return *this;
+    }
+    
+    StreamNodeIterator operator++(int)
+    {
+      StreamNodeIterator ret = *this;
+      this->operator++();
+      return ret;
+    }
+    
+    T& operator*()
+    {
+      return m_n->value;
+    }
+    
+    bool operator==(const StreamNodeIterator& rhs) const
+    {
+      return m_n == rhs.m_n;
+    }
+    
+    bool operator!=(const StreamNodeIterator& rhs) const
+    {
+      return !this->operator==(rhs);
+    }
+    
+  private:
+    StreamNode* m_sn;
+    Node* m_n;
+  };
+  
+  PinnedTally()
+    : stream_list(nullptr)
+  {
+
+  }
+  
+  PinnedTally(const PinnedTally&) = delete;
+
+
+#if defined(RAJA_ENABLE_OPENMP)
+  omp::mutex& mutex()
```

Comment:
I'd get rid of this function. using m_mutex is perfectly fine. If it's used elsewhere, there's no reason not to make it public.